stats/postgres_data_access.go

```diff
@@ -0,0 +1,186 @@
+package stats
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/intervention-engine/fhir/models"
+	_ "github.com/lib/pq"
+)
+
+// Condition coding systems
+const SnomedCodeSystem = "http://snomed.info/sct"
+const ICD9CodeSystem = "http://www.icd9data.com/"
+const ICD10CodeSystem = "http://www.icd10data.com/"
+
+var InvalidGenderError = errors.New("Invalid gender")
+var InvalidUpdateOperationError = errors.New("Invalid update operation")
+
+// StatsDataAccess is the top level interface for interacting with Synthetic Mass statistics
+type StatsDataAccess interface {
+	AddPatientStat(patient *models.Patient) error
+	RemovePatientStat(patient *models.Patient) error
+	AddConditionStat(patient *models.Patient, condition *models.Condition) error
+	RemoveConditionStat(patient *models.Patient, condition *models.Condition) error
+}
+
+// PgStatsDataAccess implements the StatsDataAccess interface using a Postgres database.
+type PgStatsDataAccess struct {
+	DB *sql.DB
+}
+
+// NewPgStatsDataAccess returns a new instance of PgStatsDataAccess.
+func NewPgStatsDataAccess(db *sql.DB) *PgStatsDataAccess {
+	return &PgStatsDataAccess{DB: db}
+}
+
+// AddPatientStat increments the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) AddPatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "increment")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// RemovePatientStat decrements the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) RemovePatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "decrement")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
```

Comment:
`AddPatientStat` and `RemovePatientStat` seem to only differ by one word -- whether `"increment"` or `"decrement"` is passed into `da.updateStats`.  As such, you could make this much DRYer by consolidating the code to a single private function that takes `"increment"` or `"decrement`" as an argument.  The public functions could just call it.

OR you could just expose one public function -- but I don't feel strongly one way or the other.  That said, if you make it one public function, then I'd say you should make the inc/dec argument more like an enum (to avoid mistakes) or you could make it an `int` called `delta` -- which generally would be called with `1` (increment) or `-1` decrement.